# Integrated project 3
Weaving the data threads of Maji Ndogo's narrative
What will we be doing?
1. Deal with some of the realities faced in many countries. 
2. Draw from different data sources to deepen the analysis into Maji 
Ndogo’s crisis.
3. Use advanced SQL tools to assemble the pieces of an audit together. 

This part of the integrated project focuses on integrating and comparing an independent auditor's report with the records in the already existing database to check for accuracy.

First, we have to load our database.

In [1]:
%reload_ext sql

In [2]:
%sql mysql+pymysql://root:Pk_0270197907@localhost:3306/md_water_services

Connecting to 'mysql+pymysql://root:***@localhost:3306/md_water_services'

To be sure we are connected to the database, let us try to retrieve records from a table in the database.

First  of all, let's create a table where we can import the auditor's report into.

In [3]:
%%sql
DROP TABLE IF EXISTS `auditor_report`;

CREATE TABLE 
    `auditor_report` (
    `location_id` VARCHAR(32),
    `type_of_water_source` VARCHAR(64),
    `true_water_source_score` int DEFAULT NULL,
    `statements` VARCHAR(255)
);

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

++
||
++
++

The table has been created and the report imported using MySQL Workbench.

In [4]:
%%sql
SELECT
    *
FROM
    employee;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

56 rows affected.

+----------------------+----------------+--------------+-------------------------------+---------------------------+---------------+-----------+----------------+
| assigned_employee_id | employee_name  | phone_number |             email             |          address          | province_name | town_name |    position    |
+----------------------+----------------+--------------+-------------------------------+---------------------------+---------------+-----------+----------------+
|          0           |  Amara Jengo   | +99637993287 |   amara.jengo@ndogowater.gov  |  36 Pwani Mchangani Road  |     Sokoto    |   Ilanga  | Field Surveyor |
|          1           |  Bello Azibo   | +99643864786 |   bello.azibo@ndogowater.gov  |   129 Ziwa La Kioo Road   |    Kilimani   |   Rural   | Field Surveyor |
|          2           |  Bakari Iniko  | +99222599041 |  bakari.iniko@ndogowater.gov  |   18 Mlima Tazama Avenue  |    Hawassa    |   Rural   | Field Surveyor |
|          3           | Malachi Mavuso | +99945849900 | malachi.mavuso@ndogowater.gov |     100 Mogadishu Road    |     Akatsi    |   Lusaka  | Field Surveyor |
|          4           |  Cheche Buhle  | +99381679640 |  cheche.buhle@ndogowater.gov  |      1 Savanna Street     |     Akatsi    |   Rural   | Field Surveyor |
|          5           | Zuriel Matembo | +99034075111 | zuriel.matembo@ndogowater.gov |  26 Bahari Ya Faraja Road |    Kilimani   |   Rural   | Field Surveyor |
|          6           |  Deka Osumare  | +99379364631 |  deka.osumare@ndogowater.gov  |    104 Kenyatta Street    |     Akatsi    |   Rural   | Field Surveyor |
|          7           | Lalitha Kaburi | +99681623240 | lalitha.kaburi@ndogowater.gov | 145 Sungura Amanpour Road |    Kilimani   |   Rural   | Field Surveyor |
|          8           |  Enitan Zuri   | +99248509202 |   enitan.zuri@ndogowater.gov  |      117 Kampala Road     |    Hawassa    |  Zanzibar | Field Surveyor |
|          10          |   Farai Nia    | +99570082739 |    farai.nia@ndogowater.gov   | 33 Angélique Kidjo Avenue |     Amanzi    |   Dahabu  | Field Surveyor |
+----------------------+----------------+--------------+-------------------------------+---------------------------+---------------+-----------+----------------+
Truncated to displaylimit of 10.

# Integrating the report
The first part of the assignmentis add the auditor's report to our database.
To do this, we need to tackle a couple of questions;
1. Is there a difference in the scores?
2. If so, are there patterns?

For the first question, we will have to compare the quality scores in the water_quality table to the auditor's scores. The auditor_report table used location_id, but the quality scores table only has a record_id we can use. The visits table links location_id and record_id, so we can link the auditor_report table and water_quality using the visits table.

So first, grab the location_id and true_water_source_score columns from auditor_report.

In [6]:
%%sql
SELECT
    location_id,
    true_water_source_score
FROM
    auditor_report
ORDER BY
    location_ID;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1620 rows affected.

+-------------+-------------------------+
| location_id | true_water_source_score |
+-------------+-------------------------+
|  AkHa00008  |            3            |
|  AkHa00053  |            9            |
|  AkHa00058  |            3            |
|  AkHa00068  |            3            |
|  AkHa00073  |            3            |
|  AkHa00088  |            1            |
|  AkHa00113  |            3            |
|  AkHa00168  |            9            |
|  AkHa00172  |            2            |
|  AkHa00193  |            9            |
+-------------+-------------------------+
Truncated to displaylimit of 10.

Now, we join the visits table to the auditor_report table. Make sure to grab subjective_quality_score, record_id and location_id.

In [7]:
%%sql
SELECT
    auditor_report.location_id AS audit_location,
    auditor_report.true_water_source_score,
    visits.location_id AS visit_location,
    visits.record_id
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

2698 rows affected.

+----------------+-------------------------+----------------+-----------+
| audit_location | true_water_source_score | visit_location | record_id |
+----------------+-------------------------+----------------+-----------+
|   SoRu34980    |            1            |   SoRu34980    |    5185   |
|   AkRu08112    |            3            |   AkRu08112    |   59367   |
|   AkLu02044    |            0            |   AkLu02044    |   37379   |
|   AkHa00421    |            3            |   AkHa00421    |   51627   |
|   SoRu35221    |            0            |   SoRu35221    |   28758   |
|   HaAm16170    |            1            |   HaAm16170    |   31048   |
|   AkRu04812    |            3            |   AkRu04812    |    1513   |
|   AkRu08304    |            3            |   AkRu08304    |    1218   |
|   AkRu05107    |            2            |   AkRu05107    |    8322   |
|   AkRu05215    |            3            |   AkRu05215    |   21160   |
+----------------+-------------------------+----------------+-----------+
Truncated to displaylimit of 10.

Now that we have the record_id for each location, our next step is to retrieve the corresponding scores from the water_quality table. We
are particularly interested in the subjective_quality_score. To do this, we'll JOIN the visits table and the water_quality table, using the
record_id as the connecting key.

In [8]:
%%sql
SELECT
    auditor_report.location_id AS audit_location,
    auditor_report.true_water_source_score,
    visits.location_id AS visit_location,
    visits.record_id,
    water_quality.subjective_quality_score
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    visits.record_id=water_quality.record_id;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

2698 rows affected.

+----------------+-------------------------+----------------+-----------+--------------------------+
| audit_location | true_water_source_score | visit_location | record_id | subjective_quality_score |
+----------------+-------------------------+----------------+-----------+--------------------------+
|   SoRu34980    |            1            |   SoRu34980    |    5185   |            1             |
|   AkRu08112    |            3            |   AkRu08112    |   59367   |            3             |
|   AkLu02044    |            0            |   AkLu02044    |   37379   |            0             |
|   AkHa00421    |            3            |   AkHa00421    |   51627   |            3             |
|   SoRu35221    |            0            |   SoRu35221    |   28758   |            0             |
|   HaAm16170    |            1            |   HaAm16170    |   31048   |            1             |
|   AkRu04812    |            3            |   AkRu04812    |    1513   |            3             |
|   AkRu08304    |            3            |   AkRu08304    |    1218   |            3             |
|   AkRu05107    |            2            |   AkRu05107    |    8322   |            2             |
|   AkRu05215    |            3            |   AkRu05215    |   21160   |            10            |
+----------------+-------------------------+----------------+-----------+--------------------------+
Truncated to displaylimit of 10.

Since it is a duplicate, we can drop one of the location_id columns. Let's leave record_id and rename the scores to surveyor_score and auditor_score to make it clear which scores we're looking at in the results set.

In [9]:
%%sql
SELECT
    auditor_report.location_id AS location_id,
    visits.record_id,
    auditor_report.true_water_source_score AS auditor_score,
    water_quality.subjective_quality_score AS surveyor_score   
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    visits.record_id=water_quality.record_id;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

2698 rows affected.

+-------------+-----------+---------------+----------------+
| location_id | record_id | auditor_score | surveyor_score |
+-------------+-----------+---------------+----------------+
|  SoRu34980  |    5185   |       1       |       1        |
|  AkRu08112  |   59367   |       3       |       3        |
|  AkLu02044  |   37379   |       0       |       0        |
|  AkHa00421  |   51627   |       3       |       3        |
|  SoRu35221  |   28758   |       0       |       0        |
|  HaAm16170  |   31048   |       1       |       1        |
|  AkRu04812  |    1513   |       3       |       3        |
|  AkRu08304  |    1218   |       3       |       3        |
|  AkRu05107  |    8322   |       2       |       2        |
|  AkRu05215  |   21160   |       3       |       10       |
+-------------+-----------+---------------+----------------+
Truncated to displaylimit of 10.

In [11]:
%config SqlMagic.displaylimit = None

displaylimit: Value None will be treated as 0 (no limit)

Now, let's try to determine if there is a difference in the scores. Note that some of sites were visited multiple times, so there can be duplicates.

In [12]:
%%sql
SELECT
    auditor_report.location_id AS location_id,
    visits.record_id,
    auditor_report.true_water_source_score AS auditor_score,
    water_quality.subjective_quality_score AS surveyor_score   
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    visits.record_id=water_quality.record_id
WHERE
    auditor_report.true_water_source_score=water_quality.subjective_quality_score
AND
    visits.visit_count=1

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1518 rows affected.

+-------------+-----------+---------------+----------------+
| location_id | record_id | auditor_score | surveyor_score |
+-------------+-----------+---------------+----------------+
|  SoRu34980  |    5185   |       1       |       1        |
|  AkRu08112  |   59367   |       3       |       3        |
|  AkLu02044  |   37379   |       0       |       0        |
|  AkHa00421  |   51627   |       3       |       3        |
|  SoRu35221  |   28758   |       0       |       0        |
|  HaAm16170  |   31048   |       1       |       1        |
|  AkRu04812  |    1513   |       3       |       3        |
|  AkRu08304  |    1218   |       3       |       3        |
|  AkRu05107  |    8322   |       2       |       2        |
|  HaDe16541  |   13070   |       2       |       2        |
|  KiRu30351  |   54220   |       9       |       9        |
|  AmAm09813  |    191    |       4       |       4        |
|  SoRu37005  |   51172   |       0       |       0        |
|  KiRu26067  |   21010   |       3       |       3        |
|  SoRu37100  |   40276   |       0       |       0        |
|  AmDa11611  |   38071   |       1       |       1        |
|  AkRu07801  |   25278   |       3       |       3        |
|  KiZu31015  |   47319   |       3       |       3        |
|  KiIs24026  |    2942   |       2       |       2        |
|  AkRu08480  |   47118   |       2       |       2        |
|  AmBe11043  |   53502   |       3       |       3        |
|  AmRu13718  |   31092   |       9       |       9        |
|  AmRu15209  |   42546   |       3       |       3        |
|  SoRu37428  |   25748   |       3       |       3        |
|  SoRu34787  |   30989   |       2       |       2        |
|  AkRu03204  |   36239   |       3       |       3        |
|  SoMa34137  |    4657   |       3       |       3        |
|  SoRu38874  |   17601   |       3       |       3        |
|  KiMr24320  |    9219   |       2       |       2        |
|  AkRu06460  |   42913   |       2       |       2        |
|  KiRu28673  |   57339   |       1       |       1        |
|  AkHa00193  |   25090   |       9       |       9        |
|  AmAs10732  |   53977   |       0       |       0        |
|  KiRu27585  |    2193   |       3       |       3        |
|  AkRu03541  |   23766   |       2       |       2        |
|  AkRu04448  |   55253   |       3       |       3        |
|  HaRu20155  |   56131   |       0       |       0        |
|  AmAm09990  |    3820   |       5       |       5        |
|  KiZu31024  |   53881   |       0       |       0        |
|  SoIl32392  |    2825   |       3       |       3        |
|  SoRu38699  |   13821   |       3       |       3        |
|  KiAm22079  |   50732   |       9       |       9        |
|  KiRu26316  |   27240   |       3       |       3        |
|  AkKi01456  |   57050   |       9       |       9        |
|  AkRu08152  |   43992   |       2       |       2        |
|  KiRu27444  |   18928   |       7       |       7        |
|  HaRu20693  |   55091   |       1       |       1        |
|  AkRu04981  |    8803   |       3       |       3        |
|  KiRu28580  |   14177   |       1       |       1        |
|  KiRu30008  |   25914   |       1       |       1        |
|  SoMa34631  |   55851   |       2       |       2        |
|  SoKo33227  |   31236   |       1       |       1        |
|  AmRu14465  |   23398   |       3       |       3        |
|  AkHa00008  |   57788   |       3       |       3        |
|  HaRu19826  |   55626   |       9       |       9        |
|  AmDa11947  |   23857   |       1       |       1        |
|  AmPw12680  |   45472   |       3       |       3        |
|  HaDe16514  |   42469   |       1       |       1        |
|  KiMr24591  |   44892   |       0       |       0        |
|  KiIs24014  |   36889   |       1       |       1        |
|  AkRu07523  |   22404   |       3       |       3        |
|  AkRu04753  |   55958   |       9       |       9        |
|  SoRu36774  |   35173   |       2

With the duplicates removed I now get 1518. What does this mean considering the auditor visited 1620 sites?

I think that is an excellent result. 1518/1620 * 100 = 94% of the records the auditor checked were correct!

But that means that 102 records are incorrect. So let's look at those. This can be done by adding one character in the last query.

In [13]:
%%sql
SELECT
    auditor_report.location_id AS location_id,
    visits.record_id,
    auditor_report.true_water_source_score AS auditor_score,
    water_quality.subjective_quality_score AS surveyor_score   
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    visits.record_id=water_quality.record_id
WHERE
    auditor_report.true_water_source_score!=water_quality.subjective_quality_score
AND
    visits.visit_count=1;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

102 rows affected.

+-------------+-----------+---------------+----------------+
| location_id | record_id | auditor_score | surveyor_score |
+-------------+-----------+---------------+----------------+
|  AkRu05215  |   21160   |       3       |       10       |
|  KiRu29290  |    7938   |       3       |       10       |
|  KiHa22748  |   43140   |       9       |       10       |
|  SoRu37841  |   18495   |       6       |       10       |
|  KiRu27884  |   33931   |       1       |       10       |
|  KiZu31170  |   17950   |       9       |       10       |
|  KiZu31370  |   36864   |       3       |       10       |
|  AkRu06495  |   45924   |       2       |       10       |
|  HaRu17528  |   30524   |       1       |       10       |
|  SoRu38331  |   13192   |       3       |       10       |
|  AmRu15556  |   15461   |       9       |       10       |
|  AmAm09607  |   55067   |       9       |       10       |
|  AkHa00314  |   29407   |       9       |       10       |
|  KiRu26598  |   28874   |       9       |       10       |
|  KiIs23853  |   55796   |       2       |       10       |
|  AkRu03475  |   33630   |       1       |       10       |
|  KiIs24083  |   28404   |       2       |       10       |
|  AmAm09631  |   14211   |       9       |       10       |
|  KiRu30343  |   51075   |       2       |       10       |
|  AmRu13668  |   10586   |       9       |       10       |
|  SoBa31691  |    3267   |       0       |       10       |
|  AkRu04116  |   53182   |       3       |       10       |
|  AmPw12272  |   50239   |       9       |       10       |
|  SoRu37524  |   59640   |       3       |       10       |
|  AmRu14691  |   40443   |       2       |       10       |
|  KiRu29556  |    5645   |       2       |       10       |
|  KiZu30847  |   57038   |       9       |       10       |
|  AkHa00363  |   25387   |       0       |       10       |
|  SoBa31761  |    3296   |       2       |       10       |
|  SoKo33094  |   16159   |       0       |       10       |
|  HaSe21323  |   26260   |       3       |       10       |
|  AmDa11466  |   17086   |       1       |       10       |
|  HaRu17038  |    7867   |       2       |       10       |
|  KiRu27364  |   38102   |       0       |       10       |
|  AkRu05880  |   12705   |       7       |       10       |
|  KiRu29696  |    251    |       2       |       10       |
|  KiRu27065  |   29772   |       0       |       10       |
|  AkRu07426  |   50918   |       3       |       10       |
|  AmAs10124  |    3154   |       6       |       10       |
|  HaRu18142  |   32555   |       2       |       10       |
|  AkRu06284  |   32008   |       3       |       10       |
|  KiRu25347  |   54560   |       2       |       10       |
|  AmRu14364  |   42302   |       2       |       10       |
|  KiZu31100  |   31033   |       3       |       10       |
|  KiRu29537  |   15869   |       2       |       10       |
|  AkRu03973  |   49033   |       3       |       10       |
|  SoKo33333  |   56898   |       2       |       10       |
|  KiRu26688  |   55794   |       2       |       10       |
|  SoIl32575  |    3116   |       1       |       10       |
|  SoMa34366  |   26183   |       9       |       10       |
|  AkRu04935  |   16410   |       3       |       10       |
|  HaRu19244  |   25725   |       2       |       10       |
|  SoRu36708  |    7252   |       3       |       10       |
|  AkRu04508  |   37357   |       1       |       10       |
|  AkRu03848  |   43412   |       3       |       10       |
|  AkLu02500  |   24797   |       3       |       10       |
|  AkRu07310  |   37457   |       9       |       10       |
|  AkKi01267  |   58866   |       9       |       10       |
|  AmRu14245  |   47227   |       9       |       10       |
|  KiZu30932  |   49082   |       9       |       10       |
|  KiAm22092  |    8186   |       9       |       10       |
|  KiRu29147  |    3179   |       0       |       10       |
|  SoRu38045  |   26566   |       9

Since we used some of this data in our previous analyses, we need to make sure those results are still valid, now we know some of them are incorrect. We didn't use the scores that much, but we relied a lot on the type_of_water_source, so let's check if there are any errors there.

In [14]:
%%sql
SELECT
    auditor_report.location_id AS location_id,
    auditor_report.type_of_water_source AS auditor_source,
    water_source.type_of_water_source AS survey_source,
    visits.record_id,
    auditor_report.true_water_source_score AS auditor_score,
    water_quality.subjective_quality_score AS surveyor_score   
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    visits.record_id = water_quality.record_id
JOIN
    water_source
ON
    visits.source_id = water_source.source_id
WHERE
    auditor_report.true_water_source_score!=water_quality.subjective_quality_score
AND
    visits.visit_count=1;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

102 rows affected.

+-------------+--------------------+--------------------+-----------+---------------+----------------+
| location_id |   auditor_source   |   survey_source    | record_id | auditor_score | surveyor_score |
+-------------+--------------------+--------------------+-----------+---------------+----------------+
|  AkRu05215  |        well        |        well        |   21160   |       3       |       10       |
|  KiRu29290  |     shared_tap     |     shared_tap     |    7938   |       3       |       10       |
|  KiHa22748  | tap_in_home_broken | tap_in_home_broken |   43140   |       9       |       10       |
|  SoRu37841  |     shared_tap     |     shared_tap     |   18495   |       6       |       10       |
|  KiRu27884  |        well        |        well        |   33931   |       1       |       10       |
|  KiZu31170  | tap_in_home_broken | tap_in_home_broken |   17950   |       9       |       10       |
|  KiZu31370  |     shared_tap     |     shared_tap     |   36864   |       3       |       10       |
|  AkRu06495  |        well        |        well        |   45924   |       2       |       10       |
|  HaRu17528  |        well        |        well        |   30524   |       1       |       10       |
|  SoRu38331  |     shared_tap     |     shared_tap     |   13192   |       3       |       10       |
|  AmRu15556  | tap_in_home_broken | tap_in_home_broken |   15461   |       9       |       10       |
|  AmAm09607  | tap_in_home_broken | tap_in_home_broken |   55067   |       9       |       10       |
|  AkHa00314  | tap_in_home_broken | tap_in_home_broken |   29407   |       9       |       10       |
|  KiRu26598  | tap_in_home_broken | tap_in_home_broken |   28874   |       9       |       10       |
|  KiIs23853  |        well        |        well        |   55796   |       2       |       10       |
|  AkRu03475  |        well        |        well        |   33630   |       1       |       10       |
|  KiIs24083  |        well        |        well        |   28404   |       2       |       10       |
|  AmAm09631  | tap_in_home_broken | tap_in_home_broken |   14211   |       9       |       10       |
|  KiRu30343  |        well        |        well        |   51075   |       2       |       10       |
|  AmRu13668  | tap_in_home_broken | tap_in_home_broken |   10586   |       9       |       10       |
|  SoBa31691  |       river        |       river        |    3267   |       0       |       10       |
|  AkRu04116  |        well        |        well        |   53182   |       3       |       10       |
|  AmPw12272  | tap_in_home_broken | tap_in_home_broken |   50239   |       9       |       10       |
|  SoRu37524  |        well        |        well        |   59640   |       3       |       10       |
|  AmRu14691  |        well        |        well        |   40443   |       2       |       10       |
|  KiRu29556  |        well        |        well        |    5645   |       2       |       10       |
|  KiZu30847  | tap_in_home_broken | tap_in_home_broken |   57038   |       9       |       10       |
|  AkHa00363  |       river        |       river        |   25387   |       0       |       10       |
|  SoBa31761  |        well        |        well        |    3296   |       2       |       10       |
|  SoKo33094  |       river        |       river        |   16159   |       0       |       10       |
|  HaSe21323  |        well        |        well        |   26260   |       3       |       10       |
|  AmDa11466  |        well        |        well        |   17086   |       1       |       10       |
|  HaRu17038  |        well        |        well        |    7867   |       2       |       10       |
|  KiRu27364  |       river        |       river        |   38102   |       0       |       10       |
|  AkRu05880  |     shared_tap     |     shared_tap     |   12705   |       7       |       10       |
|  KiRu29696  |        well        |        well        |    251    |       2       | 

So what I can see is that the types of sources look the same. So even though the scores are wrong, the integrity of the type_of_water_source data we analysed last time is not affected.

# Linking records
Linking records to employees
Next up, let's look at where these errors may have come from. At some of the locations, employees assigned scores incorrectly, and those records
ended up in this results set.

I think there are two reasons this can happen.
1. These workers are all humans and make mistakes so this is expected.
2. Unfortunately, the alternative is that someone assigned scores incorrectly on purpose.

In either case, the employees are the source of the errors, so let's JOIN the assigned_employee_id for all the people on our list from the visits table to our query. Remember, our query shows the shows the 102 incorrect records, so when we join the employee data, we can see which employees made these incorrect records.

In [15]:
%%sql
SELECT
    auditor_report.location_id AS location_id,
    visits.record_id,
    employee.assigned_employee_id,
    auditor_report.true_water_source_score AS auditor_score,
    water_quality.subjective_quality_score AS surveyor_score   
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    visits.record_id=water_quality.record_id
JOIN
    employee
ON
    visits.assigned_employee_id=employee.assigned_employee_id
WHERE
    auditor_report.true_water_source_score!=water_quality.subjective_quality_score
AND
    visits.visit_count=1;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

102 rows affected.

+-------------+-----------+----------------------+---------------+----------------+
| location_id | record_id | assigned_employee_id | auditor_score | surveyor_score |
+-------------+-----------+----------------------+---------------+----------------+
|  AkRu05215  |   21160   |          34          |       3       |       10       |
|  KiRu29290  |    7938   |          1           |       3       |       10       |
|  KiHa22748  |   43140   |          1           |       9       |       10       |
|  SoRu37841  |   18495   |          34          |       6       |       10       |
|  KiRu27884  |   33931   |          1           |       1       |       10       |
|  KiZu31170  |   17950   |          5           |       9       |       10       |
|  KiZu31370  |   36864   |          48          |       3       |       10       |
|  AkRu06495  |   45924   |          1           |       2       |       10       |
|  HaRu17528  |   30524   |          18          |       1       |       10       |
|  SoRu38331  |   13192   |          5           |       3       |       10       |
|  AmRu15556  |   15461   |          10          |       9       |       10       |
|  AmAm09607  |   55067   |          3           |       9       |       10       |
|  AkHa00314  |   29407   |          5           |       9       |       10       |
|  KiRu26598  |   28874   |          3           |       9       |       10       |
|  KiIs23853  |   55796   |          1           |       2       |       10       |
|  AkRu03475  |   33630   |          24          |       1       |       10       |
|  KiIs24083  |   28404   |          7           |       2       |       10       |
|  AmAm09631  |   14211   |          3           |       9       |       10       |
|  KiRu30343  |   51075   |          5           |       2       |       10       |
|  AmRu13668  |   10586   |          12          |       9       |       10       |
|  SoBa31691  |    3267   |          18          |       0       |       10       |
|  AkRu04116  |   53182   |          3           |       3       |       10       |
|  AmPw12272  |   50239   |          3           |       9       |       10       |
|  SoRu37524  |   59640   |          8           |       3       |       10       |
|  AmRu14691  |   40443   |          1           |       2       |       10       |
|  KiRu29556  |    5645   |          1           |       2       |       10       |
|  KiZu30847  |   57038   |          1           |       9       |       10       |
|  AkHa00363  |   25387   |          1           |       0       |       10       |
|  SoBa31761  |    3296   |          24          |       2       |       10       |
|  SoKo33094  |   16159   |          7           |       0       |       10       |
|  HaSe21323  |   26260   |          1           |       3       |       10       |
|  AmDa11466  |   17086   |          3           |       1       |       10       |
|  HaRu17038  |    7867   |          18          |       2       |       10       |
|  KiRu27364  |   38102   |          5           |       0       |       10       |
|  AkRu05880  |   12705   |          5           |       7       |       10       |
|  KiRu29696  |    251    |          1           |       2       |       10       |
|  KiRu27065  |   29772   |          1           |       0       |       10       |
|  AkRu07426  |   50918   |          34          |       3       |       10       |
|  AmAs10124  |    3154   |          10          |       6       |       10       |
|  HaRu18142  |   32555   |          38          |       2       |       10       |
|  AkRu06284  |   32008   |          34          |       3       |       10       |
|  KiRu25347  |   54560   |          3           |       2       |       10       |
|  AmRu14364  |   42302   |          1           |       2       |       10       |
|  KiZu31100  |   31033   |          1           |       3       |       10       |
|  KiRu29537  |   15869   |          24          |  

So now we can link the incorrect records to the employees who recorded them. The ID's don't help us to identify them. We have employees' names
stored along with their IDs, so let's fetch their names from the employees table instead of the ID's.

In [16]:
%%sql
SELECT
    auditor_report.location_id AS location_id,
    visits.record_id,
    employee.employee_name,
    auditor_report.true_water_source_score AS auditor_score,
    water_quality.subjective_quality_score AS surveyor_score   
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    visits.record_id=water_quality.record_id
JOIN
    employee
ON
    visits.assigned_employee_id=employee.assigned_employee_id
WHERE
    auditor_report.true_water_source_score!=water_quality.subjective_quality_score
AND
    visits.visit_count=1;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

102 rows affected.

+-------------+-----------+----------------+---------------+----------------+
| location_id | record_id | employee_name  | auditor_score | surveyor_score |
+-------------+-----------+----------------+---------------+----------------+
|  AkRu05215  |   21160   |   Rudo Imani   |       3       |       10       |
|  KiRu29290  |    7938   |  Bello Azibo   |       3       |       10       |
|  KiHa22748  |   43140   |  Bello Azibo   |       9       |       10       |
|  SoRu37841  |   18495   |   Rudo Imani   |       6       |       10       |
|  KiRu27884  |   33931   |  Bello Azibo   |       1       |       10       |
|  KiZu31170  |   17950   | Zuriel Matembo |       9       |       10       |
|  KiZu31370  |   36864   | Yewande Ebele  |       3       |       10       |
|  AkRu06495  |   45924   |  Bello Azibo   |       2       |       10       |
|  HaRu17528  |   30524   | Jengo Tumaini  |       1       |       10       |
|  SoRu38331  |   13192   | Zuriel Matembo |       3       |       10       |
|  AmRu15556  |   15461   |   Farai Nia    |       9       |       10       |
|  AmAm09607  |   55067   | Malachi Mavuso |       9       |       10       |
|  AkHa00314  |   29407   | Zuriel Matembo |       9       |       10       |
|  KiRu26598  |   28874   | Malachi Mavuso |       9       |       10       |
|  KiIs23853  |   55796   |  Bello Azibo   |       2       |       10       |
|  AkRu03475  |   33630   |  Makena Thabo  |       1       |       10       |
|  KiIs24083  |   28404   | Lalitha Kaburi |       2       |       10       |
|  AmAm09631  |   14211   | Malachi Mavuso |       9       |       10       |
|  KiRu30343  |   51075   | Zuriel Matembo |       2       |       10       |
|  AmRu13668  |   10586   |  Gamba Shani   |       9       |       10       |
|  SoBa31691  |    3267   | Jengo Tumaini  |       0       |       10       |
|  AkRu04116  |   53182   | Malachi Mavuso |       3       |       10       |
|  AmPw12272  |   50239   | Malachi Mavuso |       9       |       10       |
|  SoRu37524  |   59640   |  Enitan Zuri   |       3       |       10       |
|  AmRu14691  |   40443   |  Bello Azibo   |       2       |       10       |
|  KiRu29556  |    5645   |  Bello Azibo   |       2       |       10       |
|  KiZu30847  |   57038   |  Bello Azibo   |       9       |       10       |
|  AkHa00363  |   25387   |  Bello Azibo   |       0       |       10       |
|  SoBa31761  |    3296   |  Makena Thabo  |       2       |       10       |
|  SoKo33094  |   16159   | Lalitha Kaburi |       0       |       10       |
|  HaSe21323  |   26260   |  Bello Azibo   |       3       |       10       |
|  AmDa11466  |   17086   | Malachi Mavuso |       1       |       10       |
|  HaRu17038  |    7867   | Jengo Tumaini  |       2       |       10       |
|  KiRu27364  |   38102   | Zuriel Matembo |       0       |       10       |
|  AkRu05880  |   12705   | Zuriel Matembo |       7       |       10       |
|  KiRu29696  |    251    |  Bello Azibo   |       2       |       10       |
|  KiRu27065  |   29772   |  Bello Azibo   |       0       |       10       |
|  AkRu07426  |   50918   |   Rudo Imani   |       3       |       10       |
|  AmAs10124  |    3154   |   Farai Nia    |       6       |       10       |
|  HaRu18142  |   32555   | Thandiwe Kito  |       2       |       10       |
|  AkRu06284  |   32008   |   Rudo Imani   |       3       |       10       |
|  KiRu25347  |   54560   | Malachi Mavuso |       2       |       10       |
|  AmRu14364  |   42302   |  Bello Azibo   |       2       |       10       |
|  KiZu31100  |   31033   |  Bello Azibo   |       3       |       10       |
|  KiRu29537  |   15869   |  Makena Thabo  |       2       |       10       |
|  AkRu03973  |   49033   | Zuriel Matembo |       3       |       10       |
|  SoKo33333  |   56898   |  Bello Azibo   |       2       |       10       |
|  KiRu26688  |   55794   |   Rudo Imani   |       2       |       10       |
|  SoIl32575  |    311

In [17]:
%%sql
WITH 
    Incorrect_records AS (
        SELECT
            auditor_report.location_id AS location_id,
            visits.record_id,
            employee.employee_name,
            auditor_report.true_water_source_score AS auditor_score,
            water_quality.subjective_quality_score AS surveyor_score   
        FROM
            auditor_report
        JOIN
            visits
        ON
            auditor_report.location_id = visits.location_id
        JOIN
            water_quality
        ON
            visits.record_id=water_quality.record_id
        JOIN
            employee
        ON
            visits.assigned_employee_id=employee.assigned_employee_id
        WHERE
            auditor_report.true_water_source_score!=water_quality.subjective_quality_score
        AND
            visits.visit_count=1)
SELECT
    *
FROM
    Incorrect_records;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

102 rows affected.

+-------------+-----------+----------------+---------------+----------------+
| location_id | record_id | employee_name  | auditor_score | surveyor_score |
+-------------+-----------+----------------+---------------+----------------+
|  AkRu05215  |   21160   |   Rudo Imani   |       3       |       10       |
|  KiRu29290  |    7938   |  Bello Azibo   |       3       |       10       |
|  KiHa22748  |   43140   |  Bello Azibo   |       9       |       10       |
|  SoRu37841  |   18495   |   Rudo Imani   |       6       |       10       |
|  KiRu27884  |   33931   |  Bello Azibo   |       1       |       10       |
|  KiZu31170  |   17950   | Zuriel Matembo |       9       |       10       |
|  KiZu31370  |   36864   | Yewande Ebele  |       3       |       10       |
|  AkRu06495  |   45924   |  Bello Azibo   |       2       |       10       |
|  HaRu17528  |   30524   | Jengo Tumaini  |       1       |       10       |
|  SoRu38331  |   13192   | Zuriel Matembo |       3       |       10       |
|  AmRu15556  |   15461   |   Farai Nia    |       9       |       10       |
|  AmAm09607  |   55067   | Malachi Mavuso |       9       |       10       |
|  AkHa00314  |   29407   | Zuriel Matembo |       9       |       10       |
|  KiRu26598  |   28874   | Malachi Mavuso |       9       |       10       |
|  KiIs23853  |   55796   |  Bello Azibo   |       2       |       10       |
|  AkRu03475  |   33630   |  Makena Thabo  |       1       |       10       |
|  KiIs24083  |   28404   | Lalitha Kaburi |       2       |       10       |
|  AmAm09631  |   14211   | Malachi Mavuso |       9       |       10       |
|  KiRu30343  |   51075   | Zuriel Matembo |       2       |       10       |
|  AmRu13668  |   10586   |  Gamba Shani   |       9       |       10       |
|  SoBa31691  |    3267   | Jengo Tumaini  |       0       |       10       |
|  AkRu04116  |   53182   | Malachi Mavuso |       3       |       10       |
|  AmPw12272  |   50239   | Malachi Mavuso |       9       |       10       |
|  SoRu37524  |   59640   |  Enitan Zuri   |       3       |       10       |
|  AmRu14691  |   40443   |  Bello Azibo   |       2       |       10       |
|  KiRu29556  |    5645   |  Bello Azibo   |       2       |       10       |
|  KiZu30847  |   57038   |  Bello Azibo   |       9       |       10       |
|  AkHa00363  |   25387   |  Bello Azibo   |       0       |       10       |
|  SoBa31761  |    3296   |  Makena Thabo  |       2       |       10       |
|  SoKo33094  |   16159   | Lalitha Kaburi |       0       |       10       |
|  HaSe21323  |   26260   |  Bello Azibo   |       3       |       10       |
|  AmDa11466  |   17086   | Malachi Mavuso |       1       |       10       |
|  HaRu17038  |    7867   | Jengo Tumaini  |       2       |       10       |
|  KiRu27364  |   38102   | Zuriel Matembo |       0       |       10       |
|  AkRu05880  |   12705   | Zuriel Matembo |       7       |       10       |
|  KiRu29696  |    251    |  Bello Azibo   |       2       |       10       |
|  KiRu27065  |   29772   |  Bello Azibo   |       0       |       10       |
|  AkRu07426  |   50918   |   Rudo Imani   |       3       |       10       |
|  AmAs10124  |    3154   |   Farai Nia    |       6       |       10       |
|  HaRu18142  |   32555   | Thandiwe Kito  |       2       |       10       |
|  AkRu06284  |   32008   |   Rudo Imani   |       3       |       10       |
|  KiRu25347  |   54560   | Malachi Mavuso |       2       |       10       |
|  AmRu14364  |   42302   |  Bello Azibo   |       2       |       10       |
|  KiZu31100  |   31033   |  Bello Azibo   |       3       |       10       |
|  KiRu29537  |   15869   |  Makena Thabo  |       2       |       10       |
|  AkRu03973  |   49033   | Zuriel Matembo |       3       |       10       |
|  SoKo33333  |   56898   |  Bello Azibo   |       2       |       10       |
|  KiRu26688  |   55794   |   Rudo Imani   |       2       |       10       |
|  SoIl32575  |    311

Let's see how many unique employees are in this table.

In [18]:
%%sql
WITH 
    Incorrect_records AS (
        SELECT
            auditor_report.location_id AS location_id,
            visits.record_id,
            employee.employee_name,
            auditor_report.true_water_source_score AS auditor_score,
            water_quality.subjective_quality_score AS surveyor_score   
        FROM
            auditor_report
        JOIN
            visits
        ON
            auditor_report.location_id = visits.location_id
        JOIN
            water_quality
        ON
            visits.record_id=water_quality.record_id
        JOIN
            employee
        ON
            visits.assigned_employee_id=employee.assigned_employee_id
        WHERE
            auditor_report.true_water_source_score!=water_quality.subjective_quality_score
        AND
            visits.visit_count=1)
SELECT
    COUNT(DISTINCT employee_name) AS number_of_employees
FROM
    Incorrect_records;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

+---------------------+
| number_of_employees |
+---------------------+
|          17         |
+---------------------+

In [19]:
%%sql
WITH 
    Incorrect_records AS (
        SELECT
            auditor_report.location_id AS location_id,
            visits.record_id,
            employee.employee_name,
            auditor_report.true_water_source_score AS auditor_score,
            water_quality.subjective_quality_score AS surveyor_score   
        FROM
            auditor_report
        JOIN
            visits
        ON
            auditor_report.location_id = visits.location_id
        JOIN
            water_quality
        ON
            visits.record_id=water_quality.record_id
        JOIN
            employee
        ON
            visits.assigned_employee_id=employee.assigned_employee_id
        WHERE
            auditor_report.true_water_source_score!=water_quality.subjective_quality_score
        AND
            visits.visit_count=1)
SELECT
    employee_name,
    COUNT(employee_name) AS number_of_mistakes
FROM
    Incorrect_records
GROUP BY
    employee_name
ORDER BY
    COUNT(employee_name) DESC;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

17 rows affected.

+----------------+--------------------+
| employee_name  | number_of_mistakes |
+----------------+--------------------+
|  Bello Azibo   |         26         |
| Malachi Mavuso |         21         |
| Zuriel Matembo |         17         |
| Lalitha Kaburi |         7          |
|   Rudo Imani   |         5          |
|   Farai Nia    |         4          |
|  Enitan Zuri   |         4          |
| Yewande Ebele  |         3          |
| Jengo Tumaini  |         3          |
|  Makena Thabo  |         3          |
|  Gamba Shani   |         3          |
| Thandiwe Kito  |         1          |
|   Pili Zola    |         1          |
|   Usafi Ayo    |         1          |
|  Deka Osumare  |         1          |
|    Ona Sefu    |         1          |
|   Xola Uzuri   |         1          |
+----------------+--------------------+

# Gathering evidence
We are going to build a complex query seeking the truth.

We need to calculate the average number of mistakes employees made. We can do that by taking the average of the previous query's
results.

In [20]:
%%sql
 DROP VIEW IF EXISTS Incorrect_records;
CREATE VIEW
    Incorrect_records AS (
        SELECT
            auditor_report.location_id AS location_id,
            visits.record_id,
            employee.employee_name,
            auditor_report.true_water_source_score AS auditor_score,
            water_quality.subjective_quality_score AS surveyor_score,
            auditor_report.statements
        FROM
            auditor_report
        JOIN
            visits
        ON
            auditor_report.location_id = visits.location_id
        JOIN
            water_quality
        ON
            visits.record_id=water_quality.record_id
        JOIN
            employee
        ON
            visits.assigned_employee_id=employee.assigned_employee_id
        WHERE
            auditor_report.true_water_source_score!=water_quality.subjective_quality_score
        AND
            visits.visit_count=1)

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

++
||
++
++

In [21]:
%%sql
SELECT
    *
FROM
    incorrect_records;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

102 rows affected.

+-------------+-----------+----------------+---------------+----------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| location_id | record_id | employee_name  | auditor_score | surveyor_score |                                                                                              statements                                                                                              |
+-------------+-----------+----------------+---------------+----------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|  AkRu05215  |   21160   |   Rudo Imani   |       3       |       10       |                                               Villagers admired the official's visit for its respectful interactions, hard work, and genuine concern.                                                |
|  KiRu29290  |    7938   |  Bello Azibo   |       3       |       10       |                                           A young artist sketches the faces in the queue, capturing the weariness of daily hours spent waiting for water.                                            |
|  KiHa22748  |   43140   |  Bello Azibo   |       9       |       10       |                                                 A young girl's hopeful eyes are clouded by mistrust, her innocence tarnished by the corrupt system.                                                  |
|  SoRu37841  |   18495   |   Rudo Imani   |       6       |       10       |                                The official's respectful and diligent presence was met with heartfelt appreciation, creating a sense of closeness with the villagers.                                |
|  KiRu27884  |   33931   |  Bello Azibo   |       1       |       10       |                                                A traditional healer's empathy turns to bitterness, knowing that corrupt practices harm her community.                                                |
|  KiZu31170  |   17950   | Zuriel Matembo |       9       |       10       |                                          A community leader stood with his people, expressing concern for the water quality and the time lost in queues.","                                          |
|  KiZu31370  |   36864   | Yewande Ebele  |       3       |       10       |                                       With a keen understanding of urban challenges, the official's visit left a lasting impression of respect and commitment.                                       |
|  AkRu06495  |   45924   |  Bello Azibo   |       2       |       10       |                                             A healthcare worker in the queue expressed fears about water-borne diseases, her face etched with worry.","                                              |
|  HaRu17528  |   30524   | Jengo Tumaini  |       1       |       10       |                                         With humility and diligence, the official formed bonds with the villagers that felt like genuine family connections.                                         |
|  SoRu38331  |   13192   | Zuriel Matembo |       3       |       10       |      An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts.       |
|  AmRu15556  |   15461   |   Farai Nia    |       9       |       10       |                               The official's presence left a positive impact on the community, admired for their genuine concern and efforts to improve urban living.                                |
|  AmAm09607  |   55067   | Malachi Mavuso |       9       |       10       |            Villagers spoke of an unsettling 

In [23]:
%%sql
WITH 
    error_count AS ( -- This CTE calculates the number of mistakes each employee made
        SELECT
            employee_name,
            COUNT(employee_name) AS number_of_mistakes
        FROM
            Incorrect_records
                     
        GROUP BY
            employee_name
        ORDER BY
            number_of_mistakes DESC)
-- Query
SELECT 
    * 
FROM 
    error_count;
    
/*
Incorrect_records is a view that joins the audit report to the database
for records where the auditor and
employees scores are different*
/

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

17 rows affected.

+----------------+--------------------+
| employee_name  | number_of_mistakes |
+----------------+--------------------+
|  Bello Azibo   |         26         |
| Malachi Mavuso |         21         |
| Zuriel Matembo |         17         |
| Lalitha Kaburi |         7          |
|   Rudo Imani   |         5          |
|   Farai Nia    |         4          |
|  Enitan Zuri   |         4          |
| Yewande Ebele  |         3          |
| Jengo Tumaini  |         3          |
|  Makena Thabo  |         3          |
|  Gamba Shani   |         3          |
| Thandiwe Kito  |         1          |
|   Pili Zola    |         1          |
|   Usafi Ayo    |         1          |
|  Deka Osumare  |         1          |
|    Ona Sefu    |         1          |
|   Xola Uzuri   |         1          |
+----------------+--------------------+

In [24]:
%%sql
WITH 
    error_count AS ( -- This CTE calculates the number of mistakes each employee made
        SELECT
            employee_name,
            COUNT(employee_name) AS number_of_mistakes
        FROM
            Incorrect_records
                     
        GROUP BY
            employee_name
        ORDER BY
            number_of_mistakes DESC)
-- Query
SELECT 
    ROUND(AVG(number_of_mistakes)) AS  avg_error_count_per_empl
FROM 
    error_count;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

+--------------------------+
| avg_error_count_per_empl |
+--------------------------+
|            6             |
+--------------------------+

Finaly we have to compare each employee's error_count with avg_error_count_per_empl. We will call this results set our suspect_list.

In [25]:
%%sql
WITH 
    error_count AS ( -- This CTE calculates the number of mistakes each employee made
        SELECT
            employee_name,
            COUNT(employee_name) AS number_of_mistakes
        FROM
            Incorrect_records
                     
        GROUP BY
            employee_name
        ORDER BY
            number_of_mistakes DESC)
-- Query
SELECT
    employee_name,
    number_of_mistakes
FROM 
    error_count
WHERE
    number_of_mistakes > (SELECT 
    ROUND(AVG(number_of_mistakes)) AS  avg_error_count_per_empl
FROM 
    error_count)

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

4 rows affected.

+----------------+--------------------+
| employee_name  | number_of_mistakes |
+----------------+--------------------+
|  Bello Azibo   |         26         |
| Malachi Mavuso |         21         |
| Zuriel Matembo |         17         |
| Lalitha Kaburi |         7          |
+----------------+--------------------+

We should look at the Incorrect_records table again, and isolate all of the records these four employees gathered. We should also look at the
statements for these records to look for patterns.

In [26]:
%%sql
WITH 
    error_count AS ( 
        SELECT
            employee_name,
            COUNT(employee_name) AS number_of_mistakes
        FROM
            Incorrect_records
                     
        GROUP BY
            employee_name
        ORDER BY
            number_of_mistakes DESC),
   suspect_list AS (
    SELECT ec1.employee_name, ec1.number_of_mistakes
    FROM error_count ec1
    WHERE ec1.number_of_mistakes >= (
        SELECT AVG(ec2.number_of_mistakes)
        FROM error_count ec2
        WHERE ec2.employee_name = ec1.employee_name))
SELECT
    employee_name,
    location_id,
    statements
FROM
    Incorrect_records
WHERE
    employee_name IN (SELECT employee_name FROM suspect_list);

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

102 rows affected.

+----------------+-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| employee_name  | location_id |                                                                                              statements                                                                                              |
+----------------+-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|   Rudo Imani   |  AkRu05215  |                                               Villagers admired the official's visit for its respectful interactions, hard work, and genuine concern.                                                |
|  Bello Azibo   |  KiRu29290  |                                           A young artist sketches the faces in the queue, capturing the weariness of daily hours spent waiting for water.                                            |
|  Bello Azibo   |  KiHa22748  |                                                 A young girl's hopeful eyes are clouded by mistrust, her innocence tarnished by the corrupt system.                                                  |
|   Rudo Imani   |  SoRu37841  |                                The official's respectful and diligent presence was met with heartfelt appreciation, creating a sense of closeness with the villagers.                                |
|  Bello Azibo   |  KiRu27884  |                                                A traditional healer's empathy turns to bitterness, knowing that corrupt practices harm her community.                                                |
| Zuriel Matembo |  KiZu31170  |                                          A community leader stood with his people, expressing concern for the water quality and the time lost in queues.","                                          |
| Yewande Ebele  |  KiZu31370  |                                       With a keen understanding of urban challenges, the official's visit left a lasting impression of respect and commitment.                                       |
|  Bello Azibo   |  AkRu06495  |                                             A healthcare worker in the queue expressed fears about water-borne diseases, her face etched with worry.","                                              |
| Jengo Tumaini  |  HaRu17528  |                                         With humility and diligence, the official formed bonds with the villagers that felt like genuine family connections.                                         |
| Zuriel Matembo |  SoRu38331  |      An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts.       |
|   Farai Nia    |  AmRu15556  |                               The official's presence left a positive impact on the community, admired for their genuine concern and efforts to improve urban living.                                |
| Malachi Mavuso |  AmAm09607  |            Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.            |
| Zuriel Matembo |  AkHa00314  |                                         A street vendor's sales suffer from time spent waiting, her concern for the water's quality affecting her products.                                          |
| Malachi Mavuso |  KiRu26598  |                                           A teenager's dreams are tempered by reality, her future threatened by the corrupt practices she sees around her.                                           |
|  Bello Azibo   |  KiIs23853  |            Villagers' w

In [27]:
%%sql
DROP VIEW error_count;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

RuntimeError: (pymysql.err.OperationalError) (1051, "Unknown table 'md_water_services.error_count'")
[SQL: DROP VIEW error_count;]
(Background on this error at: https://sqlalche.me/e/20/e3q8)
If you need help solving this issue, send us a message: https://ploomber.io/community


In [28]:
%%sql
WITH 
    error_count AS ( 
        SELECT
            employee_name,
            COUNT(employee_name) AS number_of_mistakes
        FROM
            Incorrect_records
                     
        GROUP BY
            employee_name
        ORDER BY
            number_of_mistakes DESC),
    suspect_list AS (
            SELECT
                employee_name,
                number_of_mistakes
            FROM
                error_count
            WHERE
                number_of_mistakes > (SELECT 
                                        ROUND(AVG(number_of_mistakes)) AS  avg_error_count_per_empl
                                        FROM 
                                            error_count))
SELECT
    employee_name,
    location_id,
    statements
FROM
    Incorrect_records
WHERE
    employee_name IN (SELECT employee_name FROM suspect_list);

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

71 rows affected.

+----------------+-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| employee_name  | location_id |                                                                                              statements                                                                                              |
+----------------+-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|  Bello Azibo   |  KiRu29290  |                                           A young artist sketches the faces in the queue, capturing the weariness of daily hours spent waiting for water.                                            |
|  Bello Azibo   |  KiHa22748  |                                                 A young girl's hopeful eyes are clouded by mistrust, her innocence tarnished by the corrupt system.                                                  |
|  Bello Azibo   |  KiRu27884  |                                                A traditional healer's empathy turns to bitterness, knowing that corrupt practices harm her community.                                                |
| Zuriel Matembo |  KiZu31170  |                                          A community leader stood with his people, expressing concern for the water quality and the time lost in queues.","                                          |
|  Bello Azibo   |  AkRu06495  |                                             A healthcare worker in the queue expressed fears about water-borne diseases, her face etched with worry.","                                              |
| Zuriel Matembo |  SoRu38331  |      An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts.       |
| Malachi Mavuso |  AmAm09607  |            Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.            |
| Zuriel Matembo |  AkHa00314  |                                         A street vendor's sales suffer from time spent waiting, her concern for the water's quality affecting her products.                                          |
| Malachi Mavuso |  KiRu26598  |                                           A teenager's dreams are tempered by reality, her future threatened by the corrupt practices she sees around her.                                           |
|  Bello Azibo   |  KiIs23853  |            Villagers' wary accounts of an official's arrogance and detachment from their concerns raised suspicions. The mention of cash changing hands further tainted their perception.            |
| Lalitha Kaburi |  KiIs24083  |                                            A fisherman's anger boils beneath the surface, his livelihood threatened by the corrupt who tamper with water.                                            |
| Malachi Mavuso |  AmAm09631  |                                                   An elderly man lamented the endless queue, recalling times when water was abundant and clean.","                                                   |
| Zuriel Matembo |  KiRu30343  |                                            A university student's studies suffer from hours in line, dreams clouded by the daily reality of dirty water.                                             |
| Malachi Mavuso |  AkRu04116  |                                           A student's ambition is fueled by the daily water struggle, his determination to overcome evident in the queue.                                            |
| Malachi Mavuso |  AmPw12272  |                        

If you have a look, you will notice some alarming statements about these four officials (look at these records: AkRu04508, AkRu07310,
KiRu29639, AmAm09607, for example.
We want to see all statements with cash in it.

In [29]:
%%sql
WITH 
    error_count AS ( 
        SELECT
            employee_name,
            COUNT(employee_name) AS number_of_mistakes
        FROM
            Incorrect_records
                     
        GROUP BY
            employee_name
        ORDER BY
            number_of_mistakes DESC),
   suspect_list AS (
            SELECT
                employee_name,
                number_of_mistakes
            FROM
                error_count
            WHERE
                number_of_mistakes > (SELECT 
                                        ROUND(AVG(number_of_mistakes)) AS  avg_error_count_per_empl
                                        FROM 
                                            error_count))
SELECT
    employee_name,
    location_id,
    statements
FROM
    Incorrect_records
WHERE
    employee_name IN (SELECT employee_name FROM suspect_list)
AND
    statements LIKE '%cash%';

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

19 rows affected.

+----------------+-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| employee_name  | location_id |                                                                                              statements                                                                                              |
+----------------+-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Zuriel Matembo |  SoRu38331  |      An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts.       |
| Malachi Mavuso |  AmAm09607  |            Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.            |
|  Bello Azibo   |  KiIs23853  |            Villagers' wary accounts of an official's arrogance and detachment from their concerns raised suspicions. The mention of cash changing hands further tainted their perception.            |
|  Bello Azibo   |  HaSe21323  |            Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.            |
| Zuriel Matembo |  AkRu05880  |               Villagers' wary accounts of an official's arrogance and detachment from their concerns raised suspicions. The allusion to cash changing hands deepened their skepticism.               |
|  Bello Azibo   |  KiRu27065  |           Villagers expressed their discomfort with an official who displayed a haughty demeanor and negligence. The mention of cash transactions deepened their growing sense of unease.            |
| Malachi Mavuso |  KiRu25347  |              Villagers expressed their discontent with an official who appeared dismissive and neglectful. The mention of cash changing hands added to their growing sense of distrust.              |
| Zuriel Matembo |  SoIl32575  | Villagers recounted unsettling encounters with an official known for their arrogance and avoidance of responsibilities. The mention of cash changing hands added to their apprehension and distrust. |
|  Bello Azibo   |  AkRu04508  |      An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts.       |
| Lalitha Kaburi |  AkRu07310  |                Villagers spoke of their unsettling encounters with an official who seemed indifferent and uninterested, hinting at potential improprieties involving cash exchanges.                 |
|  Bello Azibo   |  KiRu29639  |          An unsettling atmosphere prevailed as villagers shared stories of an official's arrogance and perceived corruption. The mention of cash exchanges only intensified their concerns.          |
| Lalitha Kaburi |  KiRu29329  |                   Suspicion colored villagers' descriptions of an official's aloof demeanor and apparent laziness. The reference to cash transactions cast doubt on their motives.                   |
|  Bello Azibo   |  KiMr24919  |           Suspicion and unease colored the villagers' accounts of an official's haughty behavior and potential corruption. The mention of cash changing hands added to their apprehension.           |
| Zuriel Matembo |  HaSe20888  |           Suspicion and unease colored the villagers' accounts of an official's haughty behavior and potential corruption. The mention of cash changing hands added to their apprehension.           |
|  Bello Azibo   |  AmRu14887  |           Villagers exp

Now, let's see if "cash" was mentioned in statements pertaining to the other surveyors.

In [30]:
%%sql
WITH 
    error_count AS ( 
        SELECT
            employee_name,
            COUNT(employee_name) AS number_of_mistakes
        FROM
            Incorrect_records
                     
        GROUP BY
            employee_name
        ORDER BY
            number_of_mistakes DESC),
    suspect_list AS (
            SELECT
                employee_name,
                number_of_mistakes
            FROM
                error_count
            WHERE
                number_of_mistakes > (SELECT 
                                        ROUND(AVG(number_of_mistakes)) AS  avg_error_count_per_empl
                                        FROM 
                                            error_count))
SELECT
    employee_name,
    location_id,
    statements
FROM
    Incorrect_records
WHERE
    employee_name NOT IN (SELECT employee_name FROM suspect_list)
AND
    statements LIKE '%cash%';

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

+---------------+-------------+------------+
| employee_name | location_id | statements |
+---------------+-------------+------------+
+---------------+-------------+------------+

# Summary
We can see that there are 0 results returned.
So we can sum up the evidence we have for Zuriel Matembo, Malachi Mavuso, Bello Azibo and Lalitha Kaburi:
1. They all made more mistakes than their peers on average.
2. They all have incriminating statements made against them, and only them.
Keep in mind, that this is not decisive proof, but it is concerning enough that we should flag it. Pres. Naledi has worked hard to stamp out corruption, so she would urge us to report this.

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

10 rows affected.

+-------------+-----------+---------------+----------------+------------+
| location_id | record_id | auditor_score | employee_score | score_diff |
+-------------+-----------+---------------+----------------+------------+
|  SoBa31691  |    3267   |       0       |       10       |     10     |
|  AkHa00363  |   25387   |       0       |       10       |     10     |
|  SoKo33094  |   16159   |       0       |       10       |     10     |
|  KiRu27364  |   38102   |       0       |       10       |     10     |
|  KiRu27065  |   29772   |       0       |       10       |     10     |
|  KiRu29147  |    3179   |       0       |       10       |     10     |
|  SoIl32770  |    7548   |       0       |       10       |     10     |
|  KiRu29329  |   11962   |       0       |       10       |     10     |
|  SoRu39544  |   17929   |       0       |       10       |     10     |
|  SoRu38535  |   44282   |       0       |       10       |     10     |
+-------------+-----------+---------------+----------------+------------+

This last query was just to fetch other information that was not necessarily a part of the project.